packages/jss-plugin-template/src/parse.js

```diff
@@ -1,29 +1,81 @@
 // @flow
 import warning from 'tiny-warning'
 
-const semiWithNl = /;\n/
-
 /**
- * Naive CSS parser.
- * - Supports only rule body (no selectors)
- * - Requires semicolon and new line after the value (except of last line)
- * - No nested rules support
+ * A simplified CSS parser.
+ *
+ * This parser is not meant to be a complete one but to enable authoring styles
+ * using a template string with nesting syntax support, fastest parse performance and small footprint.
+ *
+ * Design of this parser has two main principles:
+ *
+ * 1. It does not parse entire CSS. It uses only specific markers to separate selectors from props and values.
+ * 2. It uses warnings to make sure expected syntax is used instead of supporting the full syntax.
+ *
+ * To do that it requires some constraints:
+ *  - Parser expects a new line after each declaration (`color: red;\n`).
+ *  - Parser expects an ampersand, selector and opening curly brace for nesting syntax on a single line (`& selector {`).
+ *  - Parser expects a closing curly brace on a separate line.
+ *
+ * Example
+ *
+ * `
+ * color: red;
+ * &:hover {
+ *   color: green;
+ * }
+ * `
  */
-export default (cssText: string) => {
+const parse = (cssText: string): Object => {
   const style = {}
-  const split = cssText.split(semiWithNl)
-  for (let i = 0; i < split.length; i++) {
-    const decl = (split[i] || '').trim()
+  const lines = cssText.split('\n')
+  const rules = [style]
+
+  for (let i = 0; i < lines.length; i++) {
+    const decl = lines[i].trim()
 
     if (!decl) continue
+
+    const ampIndex = decl.indexOf('&')
+
+    if (ampIndex !== -1) {
+      const openCurlyIndex = decl.indexOf('{')
+      if (openCurlyIndex === -1) {
+        warning(false, `[JSS] Missing opening curly brace in "${decl}".`)
+      }
+      const key = decl.substring(0, openCurlyIndex - 1)
+      const nestedStyle = {}
+      rules[rules.length - 1][key] = nestedStyle
+      rules.push(nestedStyle)
+      continue
+    }
+
+    // We are closing a nested rule.
+    if (decl === '}') {
+      rules.pop()
```

Comment:
Shouldn't we add the rule when we pop it to the style?

Reply:
we already did, we are just removing it from the array, since we basically finished adding declarations to that rule